text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,309 @@
+
+    GET /participation/v1/channels/{channelID}
+
+ or the following CLI command:
+
+    $ osnadmin channel list -c <channelID>
+
+## Guide: Prune block ledger of an existing channel
+[guide-prune-block-ledger-of-an-existing-channel]: #guide-prune-block-ledger-of-an-existing-channel
+
+To prune an existing channel ledger do the following:
+ - create a new snapshot at the desired block number,
+ - remove the OSN from the channel,
+ - join the channel from the snapshot.
+
+This approach has some drawbacks, see [Drawbacks section](#drawbacks).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Changes to the existing API
+
+This RFC adds a new endpoint to the existing **channel participation API** to allow joining a channel by snapshot. Additionally this RFC defines new **ledger management API** and introduces a new endpoint of this API to allow snapshot creation. In future we plan to add at least one more endpoint to this API which will enable ledger pruning "on the fly" without interruption in OSN service to resolve the known [drawbacks](#drawbacks).
+
+Technically we achieve the above by following these steps:
+1. Introduce new REST API endpoint **/participation/v1/joinbysnapshot** while keeping all existing functionality of the channel participation API, that is keeping "join by join-block", "remove channel", and "list channel(s)" API endpoints.
+2. Add new REST API handler for the ledger management API with prefix **/ledger/v1/**. Define a single endpoint of this API **/ledger/v1/{channelID}/snapshot** to allow snapshot creation.
+3. Add "Ledger Management API Configuration" section in the OSN YAML config file.
+4. Add support of two new snapshot operations to the CLI tool *osnadmin*, specifically the commands *osnadmin channel joinbysnapshot* and *osnadmin ledger snapshot*.
+
+## Snapshot content
+
+We propose the OSN snapshot to follow the same logical structure as peer snapshot, that is to comprise a directory with a number of files in it, one of the files being a metadata JSON file. But here the similarity ends, as the content of the files will be drastically different.
+Here is the content of the OSN snapshot:
+1. Snapshot version - to allow for the future extensions. Current version is 1.
+2. Channel ID.
+3. Last config block for the channel.
+4. lastBlockNum - the number of the last pruned block. This can be either behind the last config block, equal to the last config block number, or ahead of it.
+5. lastBlockHash - the hash of the last pruned block.
+6. prevBlockHash - technically not required, but this will make internal structure BlockchainInfo complete.
+
+Version 1 of the OSN snapshot described in this RFC defines two files as the content of the snapshot directory:
+1. Last config block for the channel. Name: "lastConfig.block"
+2. Metadata file containing all the rest of the snapshot content. Name: "metadata.json"
```

Comment:
Per RFC discussion, we need to add the snapshot block metadata here as well, so that the orderer can bootstrap for Raft processing.

Reply:
Done, here:
eda4ecc28ccb19f10faa3c7671c0cf1b2025af9c

I included only consenter-related part of metadata in the snapshot, and not all the block metadata, for the following reasons:
1. To not include the signatures
2. We have "deprecated" items in the BlockMetadata array (BlockMetadataIndex_LAST_CONFIG, BlockMetadataIndex_ORDERER), they may go away or change their semantics, so the metadata bytes produced by a newer orderer will not be identical to the same of older orderers.
3. New items may be added to the BlockMetadata array which in turn will lead again to the differing binary representations between a newer and older orderers.

The same arguments apply to the decision to remove the whole BlockMetadata from the lastConfig.block (as opposed to removing only the signatures), to keep the binary representation of it as future-proof as possible.